### PR TITLE
dm: fix validator deadlock and enhance retry

### LINKED ISSUE #9257: dm validator should retry on transient error and there should be no deadlock when bad thing happens
### What did you do?

there are 2 issues,
1. `Error 9005: Region is unavailable` is not retryable even though it could/should be transient error https://github.com/pingcap/tiflow/blob/master/dm/syncer/validate_worker.go#L170
2. there could be deadlock when errors happen, which will block any `dmctl` operation on the task, particularly, one cannot restart the task by `dmctl stop-task/start-task` because of the deadlock.

in data validator thread,

- the data validator tries to stop when encounter the first error https://github.com/pingcap/tiflow/blob/v7.1.0/dm/syncer/data_validator.go#L462
- in `stopInner`, it waits for all workers to be done https://github.com/pingcap/tiflow/blob/v7.1.0/dm/syncer/data_validator.go#L743
- thus, when it waits for the `wg`, and no one will drain the `errChan`

meantime, in worker threads,

- it reports error to the `errChan` https://github.com/pingcap/tiflow/blob/v7.1.0/dm/syncer/validate_worker.go#L170
- the bounded `errChan` will be filled fully because the validator doesn't drain the errors
- thus the workers waits on `errChan` holding the `wg` unreleased, which blocks the `innerStop` to finish

### What did you expect to see?

1. region is not available could be retried
2. there is no deadlock and the task could be restart by `stop-task/start-task` in any cases

### What did you see instead?

1. region is not available stops the validator
2. deadlock happens when a lot of workers report error

### Versions of the cluster

master

### current status of DM cluster (execute `query-status <task-name>` in dmctl)

```
{
    "result": true,
    "msg": "",
    "sources": [
        {
            "result": true,
            "msg": "",
            "sourceStatus": {
                "source": "tidbmigration-sample-src",
                "worker": "test-single-cell-dm-worker-0",
                "result": null,
                "relayStatus": null
            },
            "subTaskStatus": [
                {
                    "name": "tidbmigration-sample-task",
                    "stage": "Running",
                    "unit": "Sync",
                    "result": null,
                    "unresolvedDDLLockID": "",
                    "sync": {
                        "totalEvents": "19635499",
                        "totalTps": "41072",
                        "recentTps": "12483",
                        "masterBinlog": "(mysql-bin-changelog.061817, 57798824)",
                        "masterBinlogGtid": "",
                        "syncerBinlog": "(mysql-bin-changelog.061817, 55744546)",
                        "syncerBinlogGtid": "00000000-0000-0000-0000-000000000000:0",
                        "blockingDDLs": [
                        ],
                        "unresolvedGroups": [
                        ],
                        "synced": false,
                        "binlogType": "remote",
                        "secondsBehindMaster": "0",
                        "blockDDLOwner": "",
                        "conflictMsg": "",
                        "totalRows": "19635499",
                        "totalRps": "41072",
                        "recentRps": "12483"
                    },
                    "validation": {
                        "task": "tidbmigration-sample-task",
                        "source": "tidbmigration-sample-src",
                        "mode": "full",
                        "stage": "Running",
                        "validatorBinlog": "(mysql-bin-changelog.061762, 77247898)",
                        "validatorBinlogGtid": "00000000-0000-0000-0000-000000000000:0",
                        "result": {
                            "isCanceled": false,
                            "errors": [
                                {
                                    "ErrCode": 43005,
                                    "ErrClass": "validator",
                                    "ErrScope": "internal",
                                    "ErrLevel": "high",
                                    "Message": "failed to validate row change",
                                    "RawCause": "Error 9005: Region is unavailable",
                                    "Workaround": ""
                                }
                            ],
                            "detail": null
                        },
                        "processedRowsStatus": "insert/update/delete: 1673826/1673561/371420",
                        "pendingRowsStatus": "insert/update/delete: 17259/15936/3015",
                        "errorRowsStatus": "new/ignored/resolved: 0/0/0"
                    }
                }
            ]
        }
    ]
}
```

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: close #9257

### What is changed and how it works?
- fix deadlock between error-process-routine and worker routine when err channel is full
- remove object lock when `fillResult`
- enhance retry by reusing the resumable errors in syncer

#### Tests 

 - [x] Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

### Release note 

```release-note
dm: fix validator deadlock and enhance retry
```

### CODE REVIEW

**Comment by D3Hunter** [comment]
@lance6716 @GMHDBJD

File: dm/syncer/data_validator.go [hunk 1]
```diff
@@ -199,7 +204,7 @@ type DataValidator struct {
 
 	// fields in this field block are guarded by stateMutex
```
Comment (lance6716) [inline dm/syncer/data_validator.go:205]: seems we can update the comment about which members this mutex guards

**Review by lance6716 — COMMENTED** [review]
will review later

File: dm/unit/unit.go [hunk 2]
```diff
@@ -146,6 +147,35 @@ func IsResumableError(err *pb.ProcessError) bool {
...
+func IsResumableDBError(err error) bool {
+	if err == nil {
+		return true
+	}
+
+	err = errors.Cause(err)
+	if err == context.Canceled {
+		return false
+	}
+
```
Comment (lance6716) [inline dm/unit/unit.go:163]: as its name, should we check it's a DB error? like a error instance from mysql.MySQLError

File: dm/syncer/data_validator.go (hunk 1 above)
Comment (D3Hunter) [reply dm/syncer/data_validator.go:205]: which field do you mean? the range is not changed

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Review by lance6716 — APPROVED** [review]

**Pull request merged**